# Introduction to Pandas

Adapted by [Nimblebox Inc.](https://www.nimblebox.ai/) from [Pandas](https://pandas.pydata.org/).

## Introduction

According to the official [readme](https://github.com/pandas-dev/pandas), **pandas** is a python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive.

In this tutorial, we will cover:

* **Basics of Pandas**: Introduction Pandas Objects, creation of commonly used Pandas Objects.
* **Operations on Data**: 
* **Aggregations**: Various function used to aggregate for NumPy arrays

## Getting Started

Typically we start by importing the python package as follows.

In [1]:
import pandas as pd

## Creating data

The primary two components of pandas are the `Series` and `DataFrame`. A `Series` is essentially a column, and a `DataFrame` is a multi-dimensional table made up of a collection of `Series`.

<img width="640" height="480" src="https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png" />

DataFrames and Series are quite similar in that many operations that you can do with one you can do with the other, such as filling in null values and calculating the mean.

### DataFrame

A DataFrame is a table. It contains an array of individual entries, each of which has a certain value. Each entry corresponds to a row (or record) and a column.

For example, consider the following simple DataFrame from the `data`:

In [ ]:
data = {
    'Python': ['Guido van Rossum', '.py', 1991],
    'Javascript': ['Brendan Eich', '.js', 1995]
}

In [ ]:
programming = pd.DataFrame(data)
programming

The `Index` of this DataFrame was given to us on creation as the numbers 0-2, but we could also create our own when we initialize the DataFrame.

In [ ]:
programming = pd.DataFrame(data, index=['Creator', 'Extention', 'Year'])
programming

Now we can use `loc` to locate language's attributes based on a given Index.

In [ ]:
programming.loc['Creator']

There's more on locating and extracting data from the DataFrame which are described in the [pandas documentation]()

## Reading Data

It’s quite simple to load data from various file formats into a DataFrame. Let's see some examples of how to read data from CSV, JSON, or SQL files

### Reading data from CSVs

In [ ]:
df = pd.read_csv('programming.csv')
df

CSVs don't have indexes like our DataFrames, so all we need to do is just designate the `index_col` when reading:

In [ ]:
df = pd.read_csv('programming.csv', index_col=0)
df

Here we're setting the index to be column zero.

### Reading data from JSON

In [ ]:
df = pd.read_json('programming.json')
df

Notice this time our index came with us correctly since using JSON allowed indexes to work through nesting. Pandas will try to figure out how to create a DataFrame by analyzing structure of your JSON, and sometimes it doesn't get it right. Often you'll need to set the orient keyword argument depending on the structure, so check out [read_json docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html) about that argument to see which orientation you're using.

### Reading data from a SQL database

If you’re working with data from a SQL database you need to first establish a connection using an appropriate Python library, then pass a query to pandas. Here we'll use SQLite to demonstrate.

First, we need `pysqlite3` installed, so run this command in your terminal:

```
$ pip install pysqlite3
```


`sqlite3` is used to create a connection to a database which we can then use to generate a DataFrame through a `SELECT` query.

So first we'll make a connection to a SQLite database file:

In [ ]:
import sqlite3
con = sqlite3.connect("database.db")
df = pd.read_sql_query("SELECT * FROM programming", con)

df

Just like with CSVs, we could pass `index_col='index'`, but we can also set an index after-the-fact:

In [ ]:
df = df.set_index('index')
df

In fact, we could use `set_index()` on any DataFrame using any column at any time. Indexing Series and DataFrames is a very common task, and the different ways of doing it is worth remembering.

### Converting back to a CSV, JSON, or SQL

So after extensive work on cleaning your data, you’re now ready to save it as a file of your choice. Similar to the ways we read in data, pandas provides intuitive commands to save it:

In [ ]:
df.to_csv('programming.csv')

df.to_json('programming.json')

df.to_sql('programming', con)

When we save JSON and CSV files, all we have to input into those functions is our desired filename with the appropriate file extension. With SQL, we’re not creating a new file but instead inserting a new table into the database using our `con` variable from before.

### Most used DataFrame operations of Pandas

DataFrames possess hundreds of methods and other operations that are crucial to any kind of data analysis. Let us look at some of the most used operations that perform simple transformations of your data and those that provide fundamental statistical analysis.

Let us first load the movie dataset from a CSV and designating the movie titles to be our index:

In [23]:
movies_df = pd.read_csv("IMDB-Movie-Data.csv", index_col="Title")

#### Previewing our Datasets

The first thing to do when opening a new dataset is print out a few rows to keep as a visual reference. We accomplish this with `.head()` or by using `.tail()`:

In [24]:
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

In [25]:
movies_df.tail()

Rank                  Genre  \
Title                                                 
Secret in Their Eyes     996    Crime,Drama,Mystery   
Hostel: Part II          997                 Horror   
Step Up 2: The Streets   998    Drama,Music,Romance   
Search Party             999       Adventure,Comedy   
Nine Lives              1000  Comedy,Family,Fantasy   

                                                              Description  \
Title                                                                       
Secret in Their Eyes    A tight-knit team of rising investigators, alo...   
Hostel: Part II         Three American college students studying abroa...   
Step Up 2: The Streets  Romantic sparks occur between two dance studen...   
Search Party            A pair of friends embark on a mission to reuni...   
Nine Lives              A stuffy businessman finds himself trapped ins...   

                                Director  \
Title                                      
Secret in Their Eyes           Billy Ray   
Hostel: Part II                 Eli Roth   
Step Up 2: The Streets        Jon M. Chu   
Search Party              Scot Armstrong   
Nine Lives              Barry Sonnenfeld   

                                                                   Actors  \
Title                                                                       
Secret in Their Eyes    Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II         Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets  Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party            Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives              Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                        Year  Runtime (Minutes)  Rating  Votes  \
Title                                                            
Secret in Their Eyes    2015                111     6.2  27585   
Hostel: Part II         2007                 94     5.5  73152   
Step Up 2: The Streets  2008                 98     6.2  70699   
Search Party            2014                 93     5.6   4881   
Nine Lives              2016                 87     5.3  12435   

                        Revenue (Millions)  Metascore  
Title                                                  
Secret in Their Eyes                   NaN       45.0  
Hostel: Part II                      17.54       46.0  
Step Up 2: The Streets               58.01       50.0  
Search Party                           NaN       22.0  
Nine Lives                           19.64       11.0

By default both `.head()` and `.tail()` outputs the first and the last 5 rows respectively of our DataFrame. Typically when we load in a dataset, we like to view the first five rows to see what's under the hood like the names of each column, the index, and examples of values in each row

### Getting Info about the Data

The `.info()` operation provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using.

In [26]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                2000 non-null   int64  
 1   Genre               2000 non-null   object 
 2   Description         2000 non-null   object 
 3   Director            2000 non-null   object 
 4   Actors              2000 non-null   object 
 5   Year                2000 non-null   int64  
 6   Runtime (Minutes)   2000 non-null   int64  
 7   Rating              2000 non-null   float64
 8   Votes               2000 non-null   int64  
 9   Revenue (Millions)  1744 non-null   float64
 10  Metascore           1872 non-null   float64
dtypes: float64(3), int64(4), object(4)
memory usage: 187.5+ KB


Another useful attribute is `.shape`, which outputs just a tuple of (rows, columns).

In [27]:
movies_df.shape

(2000, 11)

### Handling Duplicates in our Dataset

It is always a good practice to verify that our dataset don't have any duplicate rows. Now we can try dropping duplicates:

In [28]:
movies_df = movies_df.drop_duplicates()

In [29]:
movies_df.shape

(1000, 11)

The `drop_duplicates()` method will return a copy of our DataFrame, and but the duplicates gets removed. With the `.shape` operation we can see that duplicate rows are removed.

### Column cleanup

Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

In [30]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

We can use the `.rename()` method to rename certain or all columns via a `dict`. Let's say we want us to rename some of the names of these columns:

In [31]:
movies_df.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue_millions'
    }, inplace=True)


movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Actors', 'Year', 'Runtime',
       'Rating', 'Votes', 'Revenue_millions', 'Metascore'],
      dtype='object')

### Working with Missing Values

When exploring data, you’ll most likely encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly you'll see Python's None or NumPy's np.nan, each of which are handled differently in some situations.

There are two options in dealing with nulls:

1. Get rid of rows or columns with nulls
2. Replace nulls with non-null values, a technique known as imputation

Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

In [32]:
movies_df.isnull()

Rank  Genre  Description  Director  Actors   Year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         Runtime  Rating  Votes  Revenue_millions  Metascore  
Title                                                                         
Guardians of the Galaxy    False   False  False             False      False  
Prometheus                 False   False  False             False      False  
Split                      False   False  False             False      False  
Sing                       False   False  False             False      False  
Suicide Squad              False   False  False             False      False  
...                          ...     ...    ...               ...        ...  
Secret in Their Eyes       False   False  False              True      False  
Hostel: Part II            False   False  False             False      False  
Step Up 2: The Streets     False   False  False             False      False  
Search Party               False   False  False              True      False  
Nine Lives                 False   False  False             False      False  

[1000 rows x 11 columns]

The function `.isnull()` just by iteself isn't very useful, and is usually used in conjunction with other methods, like `sum()`.

We can see now that our data has **128 missing** values for `revenue_millions` and **64** missing values for `metascore`.

#### Removing null values

Data Scientists and Analysts regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data.

In [34]:
movies_df.dropna()

Rank                     Genre  \
Title                                                      
Guardians of the Galaxy      1   Action,Adventure,Sci-Fi   
Prometheus                   2  Adventure,Mystery,Sci-Fi   
Split                        3           Horror,Thriller   
Sing                         4   Animation,Comedy,Family   
Suicide Squad                5  Action,Adventure,Fantasy   
...                        ...                       ...   
Resident Evil: Afterlife   994   Action,Adventure,Horror   
Project X                  995                    Comedy   
Hostel: Part II            997                    Horror   
Step Up 2: The Streets     998       Drama,Music,Romance   
Nine Lives                1000     Comedy,Family,Fantasy   

                                                                Description  \
Title                                                                         
Guardians of the Galaxy   A group of intergalactic criminals are forced ...   
Prometheus                Following clues to the origin of mankind, a te...   
Split                     Three girls are kidnapped by a man with a diag...   
Sing                      In a city of humanoid animals, a hustling thea...   
Suicide Squad             A secret government agency recruits some of th...   
...                                                                     ...   
Resident Evil: Afterlife  While still out to destroy the evil Umbrella C...   
Project X                 3 high school seniors throw a birthday party t...   
Hostel: Part II           Three American college students studying abroa...   
Step Up 2: The Streets    Romantic sparks occur between two dance studen...   
Nine Lives                A stuffy businessman finds himself trapped ins...   

                                      Director  \
Title                                            
Guardians of the Galaxy             James Gunn   
Prometheus                        Ridley Scott   
Split                       M. Night Shyamalan   
Sing                      Christophe Lourdelet   
Suicide Squad                       David Ayer   
...                                        ...   
Resident Evil: Afterlife    Paul W.S. Anderson   
Project X                      Nima Nourizadeh   
Hostel: Part II                       Eli Roth   
Step Up 2: The Streets              Jon M. Chu   
Nine Lives                    Barry Sonnenfeld   

                                                                     Actors  \
Title                                                                         
Guardians of the Galaxy   Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus                Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                     James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                      Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad             Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                     ...   
Resident Evil: Afterlife  Milla Jovovich, Ali Larter, Wentworth Miller,K...   
Project X                 Thomas Mann, Oliver Cooper, Jonathan Daniel Br...   
Hostel: Part II           Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets    Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Nine Lives                Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                          Year  Runtime  Rating   Votes  Revenue_millions  \
Title                                                                       
Guardians of the Galaxy   2014      121     8.1  757074            333.13   
Prometheus                2012      124     7.0  485820            126.46   
Split                     2016      117     7.3  157606            138.12   
Sing                      2016      108     7.2   60545            270.32   
Suicide Squad             2016      123     6.2  393727            325.02   

So in the case of our dataset, this operation would remove **128** rows where `revenue_millions` is null and **64** rows where `metascore` is null.

#### Imputation

Imputation is a conventional feature engineering technique used to keep valuable data that have null values.

There may be instances where dropping every row with a null value removes too big a chunk from your dataset, so instead we can impute that null with another value, usually the mean or the median of that column.

Let's look at imputing the missing values in the revenue_millions column. First we'll extract that column into its own variable:


In [36]:
revenue = movies_df['Revenue_millions']

Now when we select columns of a DataFrame, we use brackets just like if we were accessing a Python dictionary. `revenue` now contains a Series:

In [37]:
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: Revenue_millions, dtype: float64

We'll impute the missing values of revenue using the mean. Here's the mean value:

In [38]:
revenue_mean = revenue.mean()
revenue_mean

82.95637614678897

With the mean, let's fill the nulls using `fillna()`:

In [39]:
revenue.fillna(revenue_mean, inplace=True)

We have now replaced all nulls in `revenue` with the mean of the column. Notice that by using `inplace=True` we have actually affected the original `movies_df`:

In [40]:
movies_df.isnull().sum()

Rank                 0
Genre                0
Description          0
Director             0
Actors               0
Year                 0
Runtime              0
Rating               0
Votes                0
Revenue_millions     0
Metascore           64
dtype: int64

Imputing an entire column with the same value like this is a basic example.

#### Understanding your variables

Using `describe()` on an entire DataFrame we can get a summary of the distribution of continuous variables:

In [42]:
movies_df.describe()

Rank         Year      Runtime       Rating         Votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       Revenue_millions   Metascore  
count       1000.000000  936.000000  
mean          82.956376   58.985043  
std           96.412043   17.194757  
min            0.000000   11.000000  
25%           17.442500   47.000000  
50%           60.375000   59.500000  
75%           99.177500   72.000000  
max          936.630000  100.000000

Understanding which numbers are continuous also comes in handy when thinking about the type of plot to use to represent your data visually.

`.describe()` can also be used on a categorical variable to get the count of rows, unique count of categories, top category, and freq of top category:

In [44]:
movies_df['Genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: Genre, dtype: object

This tells us that the genre column has **207** unique values, the top value is Action/Adventure/Sci-Fi, which shows up **50** times (freq).

`.value_counts()` can tell us the frequency of all values in a column:

#### Relationships between continuous variables
By using the correlation method `.corr()` we can generate the relationship between each continuous variable:

In [45]:
movies_df.corr()

Rank      Year   Runtime    Rating     Votes  \
Rank              1.000000 -0.261605 -0.221739 -0.219555 -0.283876   
Year             -0.261605  1.000000 -0.164900 -0.211219 -0.411904   
Runtime          -0.221739 -0.164900  1.000000  0.392214  0.407062   
Rating           -0.219555 -0.211219  0.392214  1.000000  0.511537   
Votes            -0.283876 -0.411904  0.407062  0.511537  1.000000   
Revenue_millions -0.252996 -0.117562  0.247834  0.189527  0.607941   
Metascore        -0.191869 -0.079305  0.211978  0.631897  0.325684   

                  Revenue_millions  Metascore  
Rank                     -0.252996  -0.191869  
Year                     -0.117562  -0.079305  
Runtime                   0.247834   0.211978  
Rating                    0.189527   0.631897  
Votes                     0.607941   0.325684  
Revenue_millions          1.000000   0.133328  
Metascore                 0.133328   1.000000

Correlation tables are a numerical representation of the bivariate relationships in the dataset.